epochplugins/cluster.py

```diff
@@ -17,8 +21,104 @@ def populate_options(self, epoch_client: epochclient.EpochClient, subparser: arg
+        filtered_topologies = self.filter_topologies(current_cluster_data, json_data, options)
+        topologies_to_overwrite = filtered_topologies.get("topologies_to_overwrite")
+        topologies_to_create = filtered_topologies.get("topologies_to_create")
+        self.overrwrite_topologies(options, topologies_to_overwrite)
+        self.create_topologies(options, topologies_to_create)
+
+    def create_topologies(self, options, topologies_to_create):
+        if topologies_to_create:
+            print(len(topologies_to_create))
+            for data in topologies_to_create:
+                try:
+                    self.epoch_client.post("/apis/v1/topologies", data.get("topology"))
+                    print("Topology created : {topology_id}".format(topology_id=data.get("name")))
+                    self.pause_topology(data, options)
+                except Exception as ex:
+                    print("Error creating topology. Error: " + str(ex))
+
+    def pause_topology(self, data, options):
+        if options.paused:
+            self.epoch_client.put(
+                "/apis/v1/topologies/{topology_id}/pause".format(topology_id=data.get("name")),
+                None)
+
+    def overrwrite_topologies(self, options, topologies_to_overwrite):
+        if topologies_to_overwrite:
+            print(len(topologies_to_overwrite))
+            for data in topologies_to_overwrite:
+                try:
+                    self.epoch_client.put("/apis/v1/topologies", data.get("topology"))
+                    print("Topology updated : {topology_id}".format(topology_id=data.get("name")))
+                    self.pause_topology(data, options)
+                except Exception as ex:
+                    print("Error updating topology. Error: " + str(ex))
+
+    def filter_topologies(self, current_cluster_data, topologies_from_json, options: SimpleNamespace):
+        skipped_topologies = options.skip.split(',')
+        filtered_topologies = {}
+        topologies_to_overwrite = []
+        topologies_to_create = []
+        if len(skipped_topologies) > 0:
+            topologies_from_json = [topology for topology in topologies_from_json if
+                                    topology.get('id') not in skipped_topologies]
+        if not options.overwrite:
+            filtered_topologies["topologies_to_create"] = topologies_from_json
+            return filtered_topologies
+        ids = set()
+        for topology in topologies_from_json:
+            ids.add(topology.get("id"))
+        overwritten_topologies_ids = [topology for topology in current_cluster_data if (topology.get("id") in ids)]
+        for topology in topologies_from_json:
+            if topology.get("id") in overwritten_topologies_ids:
+                topologies_to_overwrite.append(topology)
+            else:
+                topologies_to_create.append(topology)
+        filtered_topologies["topologies_to_overwrite"] = topologies_to_overwrite
```

Comment:
we can create a Class object here, instead of passing along dict with keys, it'll be a more elegant way to represent the same

Reply:
class FilteredTopologies:
    def __init__(self):
        self.topologies_to_create: List[Dict[str, Any]] = []
        self.topologies_to_overwrite: List[Dict[str, Any]] = []

    def add_to_create(self, topology: Dict[str, Any]):
        self.topologies_to_create.append(topology)

    def add_to_overwrite(self, topology: Dict[str, Any]):
        self.topologies_to_overwrite.append(topology)
I will have a list of json. For json, i have to declare Dict[str, Any]] which doesn’t look nice so not sure about this

Reply:
i feel this is still better. The cleanest way would be to understand the topology entirely, but thats not necessary here.